hasher-matcher-actioner/hmalib/lambdas/actions/action_performer.py

```diff
@@ -1,33 +1,96 @@
+    This lambda is called when one or more matches are found. If a single hash matches
+    multiple datasets, this will be called only once.
+
+    Action labels are generated for each match message, then an action is performed
+    corresponding to each action label.
     """
-    for sqs_record in event["Records"]:
+    for sqs_record in event["Records"]: #confused about this... are we popping off more than one thing at a time from the "matches" queue?
         sns_notification = json.loads(sqs_record["body"])
         match_message: MatchMessage = MatchMessage.from_sns_message(
             sns_notification["Message"]
         )
+        action_labels = get_action_labels(match_message)
+        for action_label in action_labels:
+            perform_action(match_message, action_label)
+
+        # the more I write code here about reacting to ThreatExchange, the more I wish it was somewhere else entirely. Implementing
+        # here forever ties up action-related code with ThreatExchange-specific code / functionality. I prefer putting
+        # ThreatExchange-related specifics for reacting into another lambda that also subscribes to the matcher's notifications.
+        if threat_exchange_reacting_is_enabled():
+            threat_exchange_reaction_labels = get_threat_exchange_reaction_labels(action_labels)
+            if threat_exchange_reaction_labels is t.List[Label]:
+                for threat_exchange_reaction_label in threat_exchange_reaction_labels:
+                    react_to_threat_exchange(match_message, threat_exchange_reaction_label)
+
+
+def get_action_labels(match_message: MatchMessage) -> t.List[ActionLabel]:
+    """
+    TODO finish implementation
+    Returns an ActionLabel for each ActionRule that applies to a MatchMessage. 
+    """
+    action_rules = get_action_rules()
+    action_labels = [] # use Set here
+    for action_rule in action_rules:
+        if action_rule.applies(match_message) and not action_labels.contains(action_rule.action_label):
+            action_labels.append(action_rule.action_label)
+    action_labels = remove_superseded_actions(action_labels) # maybe not needed for v0
+    return action_labels
+
+
+def get_action_rules() -> t.List[ActionRule]:
+    """
+    TODO implement
+    Returns a collection of ActionRule objects. An ActionRule will have the following attributes:
+    MustHaveLabels, MustNotHaveLabels, ActionLabel
+    """
+    return [ActionRule(ActionLabel("Action", "EnqueueForReview"), [Label("Collaboration", "12345")], [])]
+
+
+def get_actions() -> t.List[Action]:
+    """
+    TODO implement
+    Returns a collection of Action objects. An Action will have the following attributes:
+    ActionLabel, Priority, SupersededByActionLabel
+    """
+    return [Action(ActionLabel("Action", "EnqueueForReview"), 1, [ActionLabel("Action", "SomeMoreImortantAction")])]
+
+
+def remove_superseded_actions(action_labels: t.List[ActionLabel]) -> t.List[ActionLabel]:
```

Comment:
The more I think about this the less I like it for v0. @Dcallies brought up a good point that the actioner is likely going to be the most sensitive part of HMA where partners are most nervous about accidentally taking the wrong or an unexpected action. If we can make the outcome as obvious as possible I think that will be best until we need more complex logic like ("dont demote if we're going to delete")

Reply:
@jeberl Yep see other comment below to @schatten